addons/binding/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/discovery/SomfyTahomaItemDiscoveryService.java

```diff
@@ -0,0 +1,277 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.somfytahoma.internal.discovery;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryServiceCallback;
+import org.eclipse.smarthome.config.discovery.ExtendedDiscoveryService;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.somfytahoma.handler.SomfyTahomaBaseThingHandler;
+import org.openhab.binding.somfytahoma.handler.SomfyTahomaBridgeHandler;
+import org.openhab.binding.somfytahoma.internal.model.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import static org.openhab.binding.somfytahoma.SomfyTahomaBindingConstants.*;
+
+/**
+ * The {@link SomfyTahomaItemDiscoveryService} discovers rollershutters and
+ * action groups associated with your TahomaLink cloud account.
+ *
+ * @author Ondrej Pecta - Initial contribution
+ */
+public class SomfyTahomaItemDiscoveryService extends AbstractDiscoveryService implements ExtendedDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(SomfyTahomaItemDiscoveryService.class);
+    private SomfyTahomaBridgeHandler bridge;
+    private DiscoveryServiceCallback discoveryServiceCallback;
+    private ScheduledFuture<?> discoveryJob;
+
+
+    private static final int DISCOVERY_TIMEOUT_SEC = 10;
+    private static final int DISCOVERY_REFRESH_SEC = 30;
+
+    public SomfyTahomaItemDiscoveryService(SomfyTahomaBridgeHandler bridgeHandler) {
+        super(DISCOVERY_TIMEOUT_SEC);
+        logger.debug("Creating discovery service");
+        this.bridge = bridgeHandler;
+    }
+
+    /**
+     * Called on component activation.
+     */
+    public void activate() {
```

Comment:
Please add `@Activate` and also take `@Nullable Map<String, @Nullable Object> configProperties` as a parameter, which potentially contains the config that should be reached through to the super implementation.

Reply:
Added

Reply:
Unfortunately I can't compile it with this parameter. I am getting 
```
Illegal redefinition of parameter configProperties, inherited method from AbstractDiscoveryService declares this parameter as '@Nullable Map<@NonNull String,@Nullable Object>' (mismatching null constraints)
```

Reply:
Then please also add the `@NonNull` to the map as it is named in the error message - then it should work.

Reply:
thanks, now it's working. Should I ```@override```  the ```activate()``` method as well as ```deactivate()```?

Reply:
Yes, please.

Reply:
> Then please also add the @NonNull to the map as it is named in the error message - then it should work.

I think this is a rather bad recommendation, because we agreed on these conventions for null annotations: https://www.openhab.org/docs/developer/development/conventions.html#null-annotations

And they state that the class should become `@NonNullByDefault` and thus in this case the non-annotated parameter will automatically become `@NonNull` without explicitly stating this.

I have spend wuite some time on removing all the `@NonNull` annotations from the source in openHAB and ESH and would not like to see newly ones added there. ESH still has some of the `@NonNull` which are formt he time before we agreed on the conventions :(

Reply:
Yeah - Stefan asked me to remove the ```@NonNull``` and now it it is back. So guys, please let me know what's the conclusion :-)

Reply:
@kaikreuzer Do you agree that we should stick to our conventions? :-)

Reply:
Well, it seems that you didn't ask for the `@NonNullByDefault` on the class in your review, otherwise it would be there, right? All I was saying is that in order to resolve the compilation error, it is required to mark the Map as NonNull. Whether this is done implicit on class level or explicit here in the code, I do not care too much.

Reply:
OK, I added ```@NonNullByDefault``` on the class and removed ```@NonNull```.